From 0e08aa60b9a4f975d6c7edbfdb14d2f94d57a57b Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:54:21 -0400
Subject: [net] sunrpc/xprtrdma: Reduce lock contention in completion handlers

Message-id: <1403798077-28799-63-git-send-email-steved@redhat.com>
Patchwork-id: 84930
O-Subject: [RHEL7.1 PATCH 62/78] xprtrmda: Reduce lock contention in completion handlers
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Chuck Lever <chuck.lever@oracle.com>

Skip the ib_poll_cq() after re-arming, if the provider knows there
are no additional items waiting. (Have a look at commit ed23a727 for
more details).

Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: 7f23f6f6e388d2003c4ecf5d558f3c2191e12530
Bugzilla: 1113248
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/verbs.c b/net/sunrpc/xprtrdma/verbs.c
index af2d097..c7d5281 100644
--- a/net/sunrpc/xprtrdma/verbs.c
+++ b/net/sunrpc/xprtrdma/verbs.c
@@ -192,8 +192,11 @@ rpcrdma_sendcq_upcall(struct ib_cq *cq, void *cq_context)
   return;
  }
 
- rc = ib_req_notify_cq(cq, IB_CQ_NEXT_COMP);
- if (rc) {
+ rc = ib_req_notify_cq(cq,
+   IB_CQ_NEXT_COMP | IB_CQ_REPORT_MISSED_EVENTS);
+ if (rc == 0)
+  return;
+ if (rc < 0) {
   dprintk("RPC:       %s: ib_req_notify_cq failed: %i\n",
    __func__, rc);
   return;
@@ -272,8 +275,11 @@ rpcrdma_recvcq_upcall(struct ib_cq *cq, void *cq_context)
   return;
  }
 
- rc = ib_req_notify_cq(cq, IB_CQ_NEXT_COMP);
- if (rc) {
+ rc = ib_req_notify_cq(cq,
+   IB_CQ_NEXT_COMP | IB_CQ_REPORT_MISSED_EVENTS);
+ if (rc == 0)
+  return;
+ if (rc < 0) {
   dprintk("RPC:       %s: ib_req_notify_cq failed: %i\n",
    __func__, rc);
   return;
-- 
1.7.1